<a href="https://colab.research.google.com/github/Fvargasv/Challenge-Telecom-X-Parte-2/blob/main/Telecom_X_parte_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Preparación de los datos

##Extracción del archivo tratado

In [1]:
import pandas as pd

datos = pd.read_csv("/content/datos_telecomx_parte2.csv")
datos.head()

customerID  Churn customer_gender  customer_SeniorCitizen  \
0  0002-ORFBO      0          Female                       0   
1  0003-MKNFE      0            Male                       0   
2  0004-TLHLJ      1            Male                       0   
3  0011-IGKFF      1            Male                       1   
4  0013-EXCHZ      1          Female                       1   

   customer_Partner  customer_Dependents  customer_tenure  phone_PhoneService  \
0                 1                    1                9                   1   
1                 0                    0                9                   1   
2                 0                    0                4                   1   
3                 1                    0               13                   1   
4                 1                    0                3                   1   

  phone_MultipleLines internet_InternetService  ... internet_DeviceProtection  \
0                  No                      DSL  ...                        No   
1                 Yes                      DSL  ...                        No   
2                  No              Fiber optic  ...                       Yes   
3                  No              Fiber optic  ...                       Yes   
4                  No              Fiber optic  ...                        No   

  internet_TechSupport internet_StreamingTV internet_StreamingMovies  \
0                  Yes                  Yes                       No   
1                   No                   No                      Yes   
2                   No                   No                       No   
3                   No                  Yes                      Yes   
4                  Yes                  Yes                       No   

         Contract PaperlessBilling     PaymentMethod  Charges_Monthly  \
0        One year                1      Mailed check             65.6   
1  Month-to-month                0      Mailed check             59.9   
2  Month-to-month                1  Electronic check             73.9   
3  Month-to-month                1  Electronic check             98.0   
4  Month-to-month                1      Mailed check             83.9   

  Charges_Total  Charges_Dayly  
0        593.30       2.186667  
1        542.40       1.996667  
2        280.85       2.463333  
3       1237.85       3.266667  
4        267.40       2.796667  

[5 rows x 22 columns]

##Eliminación de columnas irrelevantes

In [2]:
datos.columns

Index(['customerID', 'Churn', 'customer_gender', 'customer_SeniorCitizen',
       'customer_Partner', 'customer_Dependents', 'customer_tenure',
       'phone_PhoneService', 'phone_MultipleLines', 'internet_InternetService',
       'internet_OnlineSecurity', 'internet_OnlineBackup',
       'internet_DeviceProtection', 'internet_TechSupport',
       'internet_StreamingTV', 'internet_StreamingMovies', 'Contract',
       'PaperlessBilling', 'PaymentMethod', 'Charges_Monthly', 'Charges_Total',
       'Charges_Dayly'],
      dtype='object')

In [3]:
#Eliminamos la columna customerID, ya que no entrega información útil que influya en la decisión de quedarse o irse de la compañía
datos = datos.drop(['customerID'], axis=1)
datos.head()

Churn customer_gender  customer_SeniorCitizen  customer_Partner  \
0      0          Female                       0                 1   
1      0            Male                       0                 0   
2      1            Male                       0                 0   
3      1            Male                       1                 1   
4      1          Female                       1                 1   

   customer_Dependents  customer_tenure  phone_PhoneService  \
0                    1                9                   1   
1                    0                9                   1   
2                    0                4                   1   
3                    0               13                   1   
4                    0                3                   1   

  phone_MultipleLines internet_InternetService internet_OnlineSecurity  ...  \
0                  No                      DSL                      No  ...   
1                 Yes                      DSL                      No  ...   
2                  No              Fiber optic                      No  ...   
3                  No              Fiber optic                      No  ...   
4                  No              Fiber optic                      No  ...   

  internet_DeviceProtection internet_TechSupport internet_StreamingTV  \
0                        No                  Yes                  Yes   
1                        No                   No                   No   
2                       Yes                   No                   No   
3                       Yes                   No                  Yes   
4                        No                  Yes                  Yes   

  internet_StreamingMovies        Contract PaperlessBilling     PaymentMethod  \
0                       No        One year                1      Mailed check   
1                      Yes  Month-to-month                0      Mailed check   
2                       No  Month-to-month                1  Electronic check   
3                      Yes  Month-to-month                1  Electronic check   
4                       No  Month-to-month                1      Mailed check   

  Charges_Monthly  Charges_Total  Charges_Dayly  
0            65.6         593.30       2.186667  
1            59.9         542.40       1.996667  
2            73.9         280.85       2.463333  
3            98.0        1237.85       3.266667  
4            83.9         267.40       2.796667  

[5 rows x 21 columns]

##Variables explicativas y variable respuesta

In [5]:
X= datos.drop(['Churn'], axis=1)
y= datos['Churn']

In [6]:
X

customer_gender  customer_SeniorCitizen  customer_Partner  \
0             Female                       0                 1   
1               Male                       0                 0   
2               Male                       0                 0   
3               Male                       1                 1   
4             Female                       1                 1   
...              ...                     ...               ...   
7038          Female                       0                 0   
7039            Male                       0                 1   
7040            Male                       0                 0   
7041            Male                       0                 1   
7042            Male                       0                 1   

      customer_Dependents  customer_tenure  phone_PhoneService  \
0                       1                9                   1   
1                       0                9                   1   
2                       0                4                   1   
3                       0               13                   1   
4                       0                3                   1   
...                   ...              ...                 ...   
7038                    0               13                   1   
7039                    0               22                   1   
7040                    0                2                   1   
7041                    1               67                   1   
7042                    1               63                   0   

     phone_MultipleLines internet_InternetService internet_OnlineSecurity  \
0                     No                      DSL                      No   
1                    Yes                      DSL                      No   
2                     No              Fiber optic                      No   
3                     No              Fiber optic                      No   
4                     No              Fiber optic                      No   
...                  ...                      ...                     ...   
7038                  No                      DSL                     Yes   
7039                 Yes              Fiber optic                      No   
7040                  No                      DSL                      No   
7041                  No                      DSL                     Yes   
7042    No phone service                      DSL                     Yes   

     internet_OnlineBackup internet_DeviceProtection internet_TechSupport  \
0                      Yes                        No                  Yes   
1                       No                        No                   No   
2                       No                       Yes                   No   
3                      Yes                       Yes                   No   
4                       No                        No                  Yes   
...                    ...                       ...                  ...   
7038                    No                        No                  Yes   
7039                    No                        No                   No   
7040                   Yes                        No                   No   
7041                    No                       Yes                  Yes   
7042                   Yes                       Yes                   No   

     internet_StreamingTV internet_StreamingMovies        Contract  \
0                     Yes                       No        One year   
1                      No                      Yes  Month-to-month   
2                      No                       No  Month-to-month   
3                     Yes                      Yes  Month-to-month   
4                     Yes                       No  Month-to-month   
...                   ...                      ...             ...   
7038                   No                       No        One year   
7039     

In [7]:
y

0       0
1       0
2       1
3       1
4       1
       ..
7038    0
7039    1
7040    0
7041    0
7042    0
Name: Churn, Length: 7043, dtype: int64

##Encoding

Transforma las variables categóricas a formato numérico para hacerlas compatibles con los algoritmos de machine learning. Utiliza un método de codificación adecuado, como one-hot encoding.

In [4]:
from sklearn.compose import make_column_transformer
from sklearn.preprocessing import OneHotEncoder

In [8]:
columnas = X.columns

In [9]:
one_hot = make_column_transformer((OneHotEncoder(drop='if_binary'),
                                  ['customer_gender','phone_MultipleLines', 'internet_InternetService','internet_OnlineSecurity',
                                   'internet_OnlineBackup','internet_DeviceProtection','internet_TechSupport','internet_StreamingTV','internet_StreamingMovies',
                                   'Contract','PaymentMethod']),
                                  remainder='passthrough',
                                  sparse_threshold=0,
                                  force_int_remainder_cols=False)

In [10]:
X= one_hot.fit_transform(X)

In [11]:
one_hot.get_feature_names_out(columnas)

array(['onehotencoder__customer_gender_Male',
       'onehotencoder__phone_MultipleLines_No',
       'onehotencoder__phone_MultipleLines_No phone service',
       'onehotencoder__phone_MultipleLines_Yes',
       'onehotencoder__internet_InternetService_DSL',
       'onehotencoder__internet_InternetService_Fiber optic',
       'onehotencoder__internet_InternetService_No',
       'onehotencoder__internet_OnlineSecurity_No',
       'onehotencoder__internet_OnlineSecurity_No internet service',
       'onehotencoder__internet_OnlineSecurity_Yes',
       'onehotencoder__internet_OnlineBackup_No',
       'onehotencoder__internet_OnlineBackup_No internet service',
       'onehotencoder__internet_OnlineBackup_Yes',
       'onehotencoder__internet_DeviceProtection_No',
       'onehotencoder__internet_DeviceProtection_No internet service',
       'onehotencoder__internet_DeviceProtection_Yes',
       'onehotencoder__internet_TechSupport_No',
       'onehotencoder__internet_TechSupport_No internet

In [12]:
pd.DataFrame(X, columns=one_hot.get_feature_names_out(columnas)) #la transformacion no me arroja un data frame debo transformarlo en uno
#si =1 no=0

onehotencoder__customer_gender_Male  \
0                                     0.0   
1                                     1.0   
2                                     1.0   
3                                     1.0   
4                                     0.0   
...                                   ...   
7038                                  0.0   
7039                                  1.0   
7040                                  1.0   
7041                                  1.0   
7042                                  1.0   

      onehotencoder__phone_MultipleLines_No  \
0                                       1.0   
1                                       0.0   
2                                       1.0   
3                                       1.0   
4                                       1.0   
...                                     ...   
7038                                    1.0   
7039                                    0.0   
7040                                    1.0   
7041                                    1.0   
7042                                    0.0   

      onehotencoder__phone_MultipleLines_No phone service  \
0                                                   0.0     
1                                                   0.0     
2                                                   0.0     
3                                                   0.0     
4                                                   0.0     
...                                                 ...     
7038                                                0.0     
7039                                                0.0     
7040                                                0.0     
7041                                                0.0     
7042                                                1.0     

      onehotencoder__phone_MultipleLines_Yes  \
0                                        0.0   
1                                        1.0   
2                                        0.0   
3                                        0.0   
4                                        0.0   
...                                      ...   
7038                                     0.0   
7039                                     1.0   
7040                                     0.0   
7041                                     0.0   
7042                                     0.0   

      onehotencoder__internet_InternetService_DSL  \
0                                             1.0   
1                                             1.0   
2                                             0.0   
3                                             0.0   
4                                             0.0   
...                                           ...   
7038                                          1.0   
7039                                          0.0   
7040                                          1.0   
7041                                          1.0   
7042                                          1.0   

      onehotencoder__internet_InternetService_Fiber optic  \
0                                                   0.0     
1                                                   0.0     
2                                                   1.0     
3                                                   1.0     
4                                                   1.0     
...                                                 ...     
7038                                                0.0     
7039                                                1.0     
7040                                                0.0     
7041                                                0.0     
7042                                                0.0     

      onehotencoder__internet_InternetService_No  \
0                                            0.0   
1                                            0.0   
2                                            0.0   
3           

##Verificación de la Proporción de Cancelación (Churn)

In [13]:
datos['Churn'].value_counts()

Churn
0    5174
1    1869
Name: count, dtype: int64

In [17]:
datos['Churn'].value_counts(normalize=True).round(2)

Churn
0    0.73
1    0.27
Name: proportion, dtype: float64

Se ve que hay un desbalance entre la proporción de clientes que cancelaron (27%) vs los que permanecen activos (73%)

##Balanceo de Clases